File: vibecast/_discovery/eureka.py
"""HTTP/HTTPS eureka_info endpoints used by sender discovery probes."""

from __future__ import annotations

import base64
import contextlib
import hashlib
import socket
import time
from dataclasses import asdict, dataclass
from typing import TYPE_CHECKING

from aiohttp import web
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat
from cryptography.x509 import load_der_x509_certificate
from pydantic import BaseModel, ConfigDict, Field

from vibecast._log import get_logger
from vibecast._security.tls import build_server_ssl_context, load_cert_chain

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable

    from vibecast._config import EurekaDeviceCapabilitiesConfig
    from vibecast._security.certificate import CertificateBundle

log = get_logger("eureka")


class _Model(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class _Location(_Model):
    country_code: str = "US"
    latitude: float = 255.0
    longitude: float = 255.0


class _OptIn(_Model):
    crash: bool = False
    opencast: bool = False
    stats: bool = False


class _SetupStats(_Model):
    historically_succeeded: bool = True
    num_check_connectivity: int = 0
    num_connect_wifi: int = 0
    num_connected_wifi_not_saved: int = 0
    num_initial_eureka_info: int = 0
    num_obtain_ip: int = 0


class DeviceCapabilities(_Model):
    audio_hdr_supported: bool = False
    audio_surround_mode_supported: bool = False
    cast_connect_supported: bool = True
    cloud_groups_supported: bool = False
    cloudcast_supported: bool = True
    display_supported: bool = True
    fdr_supported: bool = False
    hdmi_prefer_50hz_supported: bool = False
    hdmi_prefer_high_fps_supported: bool = False
    hotspot_supported: bool = False
    https_setup_supported: bool = True
    keep_hotspot_until_connected_supported: bool = False
    multizone_supported: bool = True
    opencast_supported: bool = False
    reboot_supported: bool = False
    renaming_supported: bool = False
    set_group_audio_delay_supported: bool = False
    set_network_supported: bool = False
    setup_supported: bool = False
    stats_supported: bool = False
    system_sound_effects_supported: bool = False
    wifi_auto_save_supported: bool = False
    wifi_supported: bool = False


class EurekaDeviceInfo(_Model):
    capabilities: DeviceCapabilities
    cloud_device_id: str
    factory_country_code: str = ""
    hotspot_bssid: str = ""
    local_authorization_token_hash: str
    mac_address: str = "00:00:00:00:00:00"
    manufacturer: str = "Google Inc."
    model_name: str
    product_name: str
    public_key: str
    ssdp_udn: str
    uptime: float
    weave_device_id: str = ""


class EurekaMultizone(_Model):
    audio_output_delay: float = 0.0
    audio_output_delay_hdmi: float = 0.0
    audio_output_delay_oem: float = 0.0
    dynamic_groups: list[object] = Field(default_factory=list)
    groups: list[object] = Field(default_factory=list)
    max_static_groups: int = 100
    multichannel_status: int = 0


class EurekaInfo(_Model):
    bssid: str = ""
    build_version: str = "446070"
    cast_build_revision: str = "3.72.446070"
    connected: bool = True
    ethernet_connected: bool = True
    has_update: bool = False
    hotspot_bssid: str = ""
    ip_address: str
    locale: str = "en-US"
    location: _Location = Field(default_factory=_Location)
    mac_address: str = "00:00:00:00:00:00"
    name: str
    opt_in: _OptIn = Field(default_factory=_OptIn)
    public_key: str
    release_track: str = ""
    setup_state: int = 60
    setup_stats: _SetupStats = Field(default_factory=_SetupStats)
    ssdp_udn: str
    ssid: str = ""
    time_format: int = 1
    tos_accepted: bool = True
    uptime: float
    version: int = 12
    wpa_configured: bool = False
    wpa_state: int = 0
    device_info: EurekaDeviceInfo | None = None
    multizone: EurekaMultizone | None = None


@dataclass(slots=True, frozen=True)
class EurekaIdentity:
    friendly_name: str
    device_model: str
    ssdp_udn: str
    manufacturer: str = "Google Inc."
    locale: str = "en-US"
    country_code: str = "US"
    build_version: str = "446070"
    build_revision: str = "3.72.446070"
    capabilities: EurekaDeviceCapabilitiesConfig | None = None


class EurekaServer:
    """Expose Google Cast-style ``/setup/eureka_info`` endpoints."""

    __slots__ = (
        "_host",
        "_http_port",
        "_http_runner",
        "_http_serving_port",
        "_https_port",
        "_https_runner",
        "_https_serving_port",
        "_identity",
        "_ip_address",
        "_local_authorization_token_hash",
        "_public_key_b64",
        "_ssl_context",
        "_started_at",
    )

    def __init__(
        self,
        bundle: CertificateBundle,
        identity: EurekaIdentity,
        *,
        host: str,
        https_port: int,
        http_port: int,
    ) -> None:
        self._identity = identity
        self._host = host
        self._https_port = https_port
        self._http_port = http_port
        self._https_runner: web.AppRunner | None = None
        self._http_runner: web.AppRunner | None = None
        self._https_serving_port: int | None = None
        self._http_serving_port: int | None = None
        self._started_at = 0.0

        self._ip_address = _discover_primary_ipv4(host)
        self._public_key_b64 = _device_public_key_b64(bundle)
        self._ssl_context = build_server_ssl_context(bundle)
        self._local_authorization_token_hash = _token_hash(identity.ssdp_udn)
        log.debug(
            "eureka identity configured (name=%s model=%s ssdp_udn=%s bind_host=%s ip=%s public_key_b64_len=%d)",
            identity.friendly_name,
            identity.device_model,
            identity.ssdp_udn,
            host,
            self._ip_address,
            len(self._public_key_b64),
        )

    @property
    def https_serving_port(self) -> int | None:
        return self._https_serving_port

    @property
    def http_serving_port(self) -> int | None:
        return self._http_serving_port

    async def start(self, certificates: CertificateBundle) -> None:
        if self._https_runner is not None or self._http_runner is not None:
            log.debug("eureka server already started")
            return

        log.debug(
            "starting eureka server (host=%s, https_port=%d, http_port=%d)",
            self._host,
            self._https_port,
            self._http_port,
        )
        self._started_at = time.monotonic()
        self._public_key_b64 = _device_public_key_b64(certificates)
        self._ssl_context = build_server_ssl_context(certificates)
        ssl_context = self._ssl_context
        log.debug("eureka TLS context initialized")

        https_runner = web.AppRunner(_build_app(self._handle_eureka_info))
        await https_runner.setup()

        try:
            log.debug(
                "binding eureka HTTPS site (host=%s, port=%d)",
                self._host,
                self._https_port,
            )
            https_site = web.TCPSite(
                https_runner,
                self._host,
                self._https_port,
                ssl_context=ssl_context,
            )
            await https_site.start()
        except Exception:
            await https_runner.cleanup()
            raise

        http_runner = web.AppRunner(_build_app(self._handle_eureka_info))
        await http_runner.setup()

        try:
            log.debug(
                "binding eureka HTTP site (host=%s, port=%d)",
                self._host,
                self._http_port,
            )
            http_site = web.TCPSite(http_runner, self._host, self._http_port)
            await http_site.start()
        except Exception:
            await http_runner.cleanup()
            await https_runner.cleanup()
            raise

        self._https_serving_port = (
            _resolve_serving_port(https_runner) or self._https_port
        )
        self._http_serving_port = _resolve_serving_port(http_runner) or self._http_port
        self._https_runner = https_runner
        self._http_runner = http_runner
        log.info(
            "eureka server started (host=%s, https=%d, http=%d)",
            self._host,
            self._https_serving_port,
            self._http_serving_port,
        )

    def update_certificate(self, bundle: CertificateBundle) -> None:
        """Hot-reload certificate material for future HTTPS handshakes."""
        load_cert_chain(self._ssl_context, bundle)
        self._public_key_b64 = _device_public_key_b64(bundle)
        log.info(
            "eureka TLS certificate rotated (valid=%s -> %s)",
            bundle.not_valid_before.isoformat(),
            bundle.not_valid_after.isoformat(),
        )

    async def stop(self) -> None:
        http_runner = self._http_runner
        https_runner = self._https_runner
        self._http_runner = None
        self._https_runner = None

        if http_runner is None and https_runner is None:
            log.debug("eureka server already stopped")
            return

        with contextlib.suppress(Exception):
            if http_runner is not None:
                await http_runner.cleanup()
        with contextlib.suppress(Exception):
            if https_runner is not None:
                await https_runner.cleanup()

        self._http_serving_port = None
        self._https_serving_port = None
        self._started_at = 0.0
        log.info("eureka server stopped")

    async def _handle_eureka_info(self, request: web.Request) -> web.Response:
        raw_params = request.query.get("params")
        params = _parse_params(raw_params)
        payload = self._build_payload(params)
        log.debug(
            "served /setup/eureka_info (remote=%s scheme=%s params=%r keys=%d)",
            request.remote,
            request.scheme,
            raw_params,
            len(payload),
        )
        return web.json_response(payload)

    def _build_payload(self, params: tuple[str, ...] | None) -> dict[str, object]:
        uptime = max(time.monotonic() - self._started_at, 0.0)
        product_name = _product_name(self._identity.device_model)
        cloud_device_id = _cloud_device_id(self._identity.ssdp_udn)
        capabilities = (
            DeviceCapabilities(**asdict(self._identity.capabilities))
            if self._identity.capabilities is not None
            else DeviceCapabilities()
        )

        full_response = EurekaInfo(
            build_version=self._identity.build_version,
            cast_build_revision=self._identity.build_revision,
            ip_address=self._ip_address,
            locale=self._identity.locale,
            location=_Location(country_code=self._identity.country_code),
            name=self._identity.friendly_name,
            public_key=self._public_key_b64,
            ssdp_udn=self._identity.ssdp_udn,
            uptime=uptime,
            device_info=EurekaDeviceInfo(
                capabilities=capabilities,
                cloud_device_id=cloud_device_id,
                local_authorization_token_hash=self._local_authorization_token_hash,
                manufacturer=self._identity.manufacturer,
                model_name=self._identity.device_model,
                product_name=product_name,
                public_key=self._public_key_b64,
                ssdp_udn=self._identity.ssdp_udn,
                uptime=uptime,
            ),
            multizone=EurekaMultizone(),
        )

        encoded = full_response.model_dump(exclude_none=True)
        if params is None:
            # Match real devices: the non-filtered payload does not include these
            # optional blocks unless explicitly requested via ``?params=...``.
            _ = encoded.pop("device_info", None)
            _ = encoded.pop("multizone", None)
            return encoded

        return {key: encoded[key] for key in params if key in encoded}


def _build_app(
    handler: Callable[[web.Request], Awaitable[web.StreamResponse]],
) -> web.Application:
    app = web.Application()
    _ = app.router.add_get("/setup/eureka_info", handler)
    return app


def _resolve_serving_port(runner: web.AppRunner) -> int | None:
    addresses = runner.addresses
    if not addresses:
        return None

    host, port, *_ = addresses[0]
    _ = host
    return int(port)


def _discover_primary_ipv4(bind_host: str) -> str:
    if bind_host not in {"", "0.0.0.0", "::"}:
        log.debug("using explicit bind host as eureka IP: %s", bind_host)
        return bind_host

    try:
        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
            sock.connect(("224.0.0.251", 5353))
            discovered = str(sock.getsockname()[0])
            log.debug(
                "discovered eureka primary IPv4 via multicast probe: %s", discovered
            )
            return discovered
    except OSError:
        log.debug(
            "failed to discover eureka primary IPv4 via multicast probe", exc_info=True
        )

    log.debug("falling back to loopback eureka IP: 127.0.0.1")
    return "127.0.0.1"


def _parse_params(raw: str | None) -> tuple[str, ...] | None:
    if raw is None:
        return None
    values = tuple(item.strip() for item in raw.split(",") if item.strip())
    return values if values else None


def _cloud_device_id(ssdp_udn: str) -> str:
    cleaned = ssdp_udn.replace("-", "").upper()
    if len(cleaned) == 32 and cleaned.isalnum():
        return cleaned
    return hashlib.md5(ssdp_udn.encode("utf-8")).hexdigest().upper()  # noqa: S324


def _product_name(model_name: str) -> str:
    normalized = "".join(ch.lower() for ch in model_name if ch.isalnum())
    return normalized or "chromecast"


def _token_hash(ssdp_udn: str) -> str:
    digest = hashlib.sha256(ssdp_udn.encode("utf-8")).digest()
    return base64.b64encode(digest).decode("ascii")


def _device_public_key_b64(bundle: CertificateBundle) -> str:
    cert = load_der_x509_certificate(bundle.device_cert_der)
    pub = cert.public_key().public_bytes(
        encoding=Encoding.DER,
        format=PublicFormat.SubjectPublicKeyInfo,
    )
    return base64.b64encode(pub).decode("ascii")


__all__ = [
    "DeviceCapabilities",
    "EurekaDeviceInfo",
    "EurekaIdentity",
    "EurekaInfo",
    "EurekaMultizone",
    "EurekaServer",
]

File: crates/vibecast-discovery/src/eureka.rs
//! `/setup/eureka_info` endpoints probed by Cast senders during discovery.
//!
//! The JSON payload mirrors a real Chromecast. The same handler serves HTTP
//! (8008) and HTTPS (8443); the payload is byte-for-byte identical modulo key
//! ordering, which senders don't depend on.

use std::sync::Arc;
use std::time::Instant;

use axum::extract::{Query, State};
use axum::routing::get;
use axum::{Json, Router};
use base64::prelude::{Engine, BASE64_STANDARD};
use md5::{Digest, Md5};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use sha2::Sha256;

use vibecast_security::CertificateBundle;

use crate::error::DiscoveryError;

/// Device capability flags advertised in eureka `device_info.capabilities`.
///
/// Also deserialized directly from the `[device.capabilities]` config table
/// (missing keys fall back to the Chromecast-like defaults).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(default, deny_unknown_fields)]
#[allow(missing_docs)]
pub struct DeviceCapabilities {
    pub audio_hdr_supported: bool,
    pub audio_surround_mode_supported: bool,
    pub cast_connect_supported: bool,
    pub cloud_groups_supported: bool,
    pub cloudcast_supported: bool,
    pub display_supported: bool,
    pub fdr_supported: bool,
    pub hdmi_prefer_50hz_supported: bool,
    pub hdmi_prefer_high_fps_supported: bool,
    pub hotspot_supported: bool,
    pub https_setup_supported: bool,
    pub keep_hotspot_until_connected_supported: bool,
    pub multizone_supported: bool,
    pub opencast_supported: bool,
    pub reboot_supported: bool,
    pub renaming_supported: bool,
    pub set_group_audio_delay_supported: bool,
    pub set_network_supported: bool,
    pub setup_supported: bool,
    pub stats_supported: bool,
    pub system_sound_effects_supported: bool,
    pub wifi_auto_save_supported: bool,
    pub wifi_supported: bool,
}

impl Default for DeviceCapabilities {
    fn default() -> Self {
        Self {
            audio_hdr_supported: false,
            audio_surround_mode_supported: false,
            cast_connect_supported: true,
            cloud_groups_supported: false,
            cloudcast_supported: true,
            display_supported: true,
            fdr_supported: false,
            hdmi_prefer_50hz_supported: false,
            hdmi_prefer_high_fps_supported: false,
            hotspot_supported: false,
            https_setup_supported: true,
            keep_hotspot_until_connected_supported: false,
            multizone_supported: true,
            opencast_supported: false,
            reboot_supported: false,
            renaming_supported: false,
            set_group_audio_delay_supported: false,
            set_network_supported: false,
            setup_supported: false,
            stats_supported: false,
            system_sound_effects_supported: false,
            wifi_auto_save_supported: false,
            wifi_supported: false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
struct Location {
    country_code: String,
    latitude: f64,
    longitude: f64,
}

#[derive(Debug, Clone, Default, Serialize)]
struct OptIn {
    crash: bool,
    opencast: bool,
    stats: bool,
}

#[derive(Debug, Clone, Serialize)]
struct SetupStats {
    historically_succeeded: bool,
    num_check_connectivity: u32,
    num_connect_wifi: u32,
    num_connected_wifi_not_saved: u32,
    num_initial_eureka_info: u32,
    num_obtain_ip: u32,
}

impl Default for SetupStats {
    fn default() -> Self {
        Self {
            historically_succeeded: true,
            num_check_connectivity: 0,
            num_connect_wifi: 0,
            num_connected_wifi_not_saved: 0,
            num_initial_eureka_info: 0,
            num_obtain_ip: 0,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
struct DeviceInfo {
    capabilities: DeviceCapabilities,
    cloud_device_id: String,
    factory_country_code: String,
    hotspot_bssid: String,
    local_authorization_token_hash: String,
    mac_address: String,
    manufacturer: String,
    model_name: String,
    product_name: String,
    public_key: String,
    ssdp_udn: String,
    uptime: f64,
    weave_device_id: String,
}

#[derive(Debug, Clone, Serialize)]
struct Multizone {
    audio_output_delay: f64,
    audio_output_delay_hdmi: f64,
    audio_output_delay_oem: f64,
    dynamic_groups: Vec<Value>,
    groups: Vec<Value>,
    max_static_groups: u32,
    multichannel_status: u32,
}

impl Default for Multizone {
    fn default() -> Self {
        Self {
            audio_output_delay: 0.0,
            audio_output_delay_hdmi: 0.0,
            audio_output_delay_oem: 0.0,
            dynamic_groups: Vec::new(),
            groups: Vec::new(),
            max_static_groups: 100,
            multichannel_status: 0,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
struct EurekaInfo {
    bssid: String,
    build_version: String,
    cast_build_revision: String,
    connected: bool,
    ethernet_connected: bool,
    has_update: bool,
    hotspot_bssid: String,
    ip_address: String,
    locale: String,
    location: Location,
    mac_address: String,
    name: String,
    opt_in: OptIn,
    public_key: String,
    release_track: String,
    setup_state: u32,
    setup_stats: SetupStats,
    ssdp_udn: String,
    ssid: String,
    time_format: u32,
    tos_accepted: bool,
    uptime: f64,
    version: u32,
    wpa_configured: bool,
    wpa_state: u32,
    device_info: DeviceInfo,
    multizone: Multizone,
}

/// Static and per-device identity for the eureka payload.
#[derive(Debug, Clone)]
pub struct EurekaIdentity {
    /// Friendly device name.
    pub friendly_name: String,
    /// Device model string.
    pub device_model: String,
    /// SSDP UDN (also used for cloud id and token hash).
    pub ssdp_udn: String,
    /// The device's advertised IP address.
    pub ip_address: String,
    /// Manufacturer (default "Google Inc.").
    pub manufacturer: String,
    /// Locale (default "en-US").
    pub locale: String,
    /// Country code (default "US").
    pub country_code: String,
    /// Build version string.
    pub build_version: String,
    /// Cast build revision string.
    pub build_revision: String,
    /// Optional capability overrides.
    pub capabilities: Option<DeviceCapabilities>,
}

impl EurekaIdentity {
    /// Build an identity with Cast-default fields.
    #[must_use]
    pub fn new(
        friendly_name: String,
        device_model: String,
        ssdp_udn: String,
        ip_address: String,
    ) -> Self {
        Self {
            friendly_name,
            device_model,
            ssdp_udn,
            ip_address,
            manufacturer: "Google Inc.".into(),
            locale: "en-US".into(),
            country_code: "US".into(),
            build_version: "446070".into(),
            build_revision: "3.72.446070".into(),
            capabilities: None,
        }
    }
}

struct EurekaState {
    identity: EurekaIdentity,
    public_key_b64: String,
    token_hash: String,
    started: Instant,
}

/// Serves the eureka discovery endpoint over HTTP and/or HTTPS.
pub struct EurekaServer {
    state: Arc<EurekaState>,
}

impl EurekaServer {
    /// Build the server from certificate material (for the device public key)
    /// and identity.
    pub fn new(
        bundle: &CertificateBundle,
        identity: EurekaIdentity,
    ) -> Result<Self, DiscoveryError> {
        let public_key_b64 = BASE64_STANDARD.encode(bundle.device_public_key_der()?);
        let token_hash = token_hash(&identity.ssdp_udn);
        Ok(Self {
            state: Arc::new(EurekaState {
                identity,
                public_key_b64,
                token_hash,
                started: Instant::now(),
            }),
        })
    }

    /// The axum router exposing `/setup/eureka_info`.
    pub fn router(&self) -> Router {
        Router::new()
            .route("/setup/eureka_info", get(handle_eureka_info))
            .with_state(Arc::clone(&self.state))
    }

    /// Serve HTTP on the given listener until it errors.
    pub async fn serve_http(
        &self,
        listener: tokio::net::TcpListener,
    ) -> Result<(), DiscoveryError> {
        axum::serve(listener, self.router()).await?;
        Ok(())
    }

    /// Serve HTTPS on the given std listener using `config` for TLS.
    pub async fn serve_https(
        &self,
        listener: std::net::TcpListener,
        config: rustls::ServerConfig,
    ) -> Result<(), DiscoveryError> {
        let tls = axum_server::tls_rustls::RustlsConfig::from_config(Arc::new(config));
        axum_server::from_tcp_rustls(listener, tls)?
            .serve(self.router().into_make_service())
            .await?;
        Ok(())
    }
}

#[derive(Deserialize)]
struct ParamsQuery {
    params: Option<String>,
}

async fn handle_eureka_info(
    State(state): State<Arc<EurekaState>>,
    Query(query): Query<ParamsQuery>,
) -> Json<Value> {
    let params = query.params.as_deref().and_then(parse_params);
    Json(build_payload(&state, params.as_deref()))
}

fn parse_params(raw: &str) -> Option<Vec<String>> {
    let values: Vec<String> = raw
        .split(',')
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .map(str::to_owned)
        .collect();
    (!values.is_empty()).then_some(values)
}

fn build_payload(state: &EurekaState, params: Option<&[String]>) -> Value {
    let uptime = state.started.elapsed().as_secs_f64();
    let id = &state.identity;
    let product_name = product_name(&id.device_model);
    let cloud_device_id = cloud_device_id(&id.ssdp_udn);

    let info = EurekaInfo {
        bssid: String::new(),
        build_version: id.build_version.clone(),
        cast_build_revision: id.build_revision.clone(),
        connected: true,
        ethernet_connected: true,
        has_update: false,
        hotspot_bssid: String::new(),
        ip_address: id.ip_address.clone(),
        locale: id.locale.clone(),
        location: Location {
            country_code: id.country_code.clone(),
            latitude: 255.0,
            longitude: 255.0,
        },
        mac_address: "00:00:00:00:00:00".into(),
        name: id.friendly_name.clone(),
        opt_in: OptIn::default(),
        public_key: state.public_key_b64.clone(),
        release_track: String::new(),
        setup_state: 60,
        setup_stats: SetupStats::default(),
        ssdp_udn: id.ssdp_udn.clone(),
        ssid: String::new(),
        time_format: 1,
        tos_accepted: true,
        uptime,
        version: 12,
        wpa_configured: false,
        wpa_state: 0,
        device_info: DeviceInfo {
            capabilities: id.capabilities.clone().unwrap_or_default(),
            cloud_device_id,
            factory_country_code: String::new(),
            hotspot_bssid: String::new(),
            local_authorization_token_hash: state.token_hash.clone(),
            mac_address: "00:00:00:00:00:00".into(),
            manufacturer: id.manufacturer.clone(),
            model_name: id.device_model.clone(),
            product_name,
            public_key: state.public_key_b64.clone(),
            ssdp_udn: id.ssdp_udn.clone(),
            uptime,
            weave_device_id: String::new(),
        },
        multizone: Multizone::default(),
    };

    let mut value = serde_json::to_value(&info).expect("eureka info serializes");
    let object = value.as_object_mut().expect("eureka info is an object");

    match params {
        // Unfiltered response omits the optional blocks, like a real device.
        None => {
            object.remove("device_info");
            object.remove("multizone");
            value
        }
        Some(keys) => {
            let filtered: Map<String, Value> = keys
                .iter()
                .filter_map(|key| object.get(key).map(|v| (key.clone(), v.clone())))
                .collect();
            Value::Object(filtered)
        }
    }
}

fn cloud_device_id(ssdp_udn: &str) -> String {
    let cleaned: String = ssdp_udn.replace('-', "").to_uppercase();
    if cleaned.len() == 32 && cleaned.chars().all(|c| c.is_ascii_alphanumeric()) {
        cleaned
    } else {
        hex::encode_upper(Md5::digest(ssdp_udn.as_bytes()))
    }
}

fn product_name(model_name: &str) -> String {
    let normalized: String = model_name
        .chars()
        .filter(|c| c.is_ascii_alphanumeric())
        .map(|c| c.to_ascii_lowercase())
        .collect();
    if normalized.is_empty() {
        "chromecast".into()
    } else {
        normalized
    }
}

fn token_hash(ssdp_udn: &str) -> String {
    BASE64_STANDARD.encode(Sha256::digest(ssdp_udn.as_bytes()))
}

#[cfg(test)]
mod tests {
    use axum::body::{to_bytes, Body};
    use axum::http::{Request, StatusCode};
    use tower::ServiceExt;

    use super::*;

    fn bundle() -> CertificateBundle {
        let key = rcgen::KeyPair::generate().unwrap();
        let cert = rcgen::CertificateParams::new(vec!["Device".to_string()])
            .unwrap()
            .self_signed(&key)
            .unwrap();
        CertificateBundle {
            peer_cert_pem: Vec::new(),
            peer_key_pem: Vec::new(),
            peer_cert_der: cert.der().to_vec(),
            device_cert_der: cert.der().to_vec(),
            intermediate_certs_der: Vec::new(),
            signature_sha1: Vec::new(),
            signature_sha256: Vec::new(),
            not_valid_before: 0,
            not_valid_after: i64::MAX,
            crl: None,
        }
    }

    fn server() -> EurekaServer {
        let identity = EurekaIdentity::new(
            "Living Room".into(),
            "Chromecast".into(),
            "12345678-1234-1234-1234-123456789abc".into(),
            "192.168.1.42".into(),
        );
        EurekaServer::new(&bundle(), identity).unwrap()
    }

    async fn get(uri: &str) -> Value {
        let response = server()
            .router()
            .oneshot(Request::builder().uri(uri).body(Body::empty()).unwrap())
            .await
            .unwrap();
        assert_eq!(response.status(), StatusCode::OK);
        let bytes = to_bytes(response.into_body(), usize::MAX).await.unwrap();
        serde_json::from_slice(&bytes).unwrap()
    }

    #[tokio::test]
    async fn unfiltered_response_omits_optional_blocks() {
        let body = get("/setup/eureka_info").await;
        assert_eq!(body["name"], "Living Room");
        assert_eq!(body["ip_address"], "192.168.1.42");
        assert_eq!(body["version"], 12);
        assert_eq!(body["build_version"], "446070");
        assert_eq!(body["ssdp_udn"], "12345678-1234-1234-1234-123456789abc");
        assert!(!body["public_key"].as_str().unwrap().is_empty());
        // Optional blocks are omitted unless requested.
        assert!(body.get("device_info").is_none());
        assert!(body.get("multizone").is_none());
    }

    #[tokio::test]
    async fn params_filters_to_requested_keys_including_device_info() {
        let body = get("/setup/eureka_info?params=name,version,device_info,multizone").await;
        let object = body.as_object().unwrap();
        assert_eq!(object.len(), 4);
        assert_eq!(body["name"], "Living Room");
        assert_eq!(body["version"], 12);

        let device_info = &body["device_info"];
        assert_eq!(device_info["model_name"], "Chromecast");
        assert_eq!(device_info["product_name"], "chromecast");
        assert_eq!(device_info["manufacturer"], "Google Inc.");
        // token hash is base64 of sha256(ssdp_udn)
        assert!(!device_info["local_authorization_token_hash"]
            .as_str()
            .unwrap()
            .is_empty());
        assert_eq!(device_info["capabilities"]["cast_connect_supported"], true);

        assert_eq!(body["multizone"]["max_static_groups"], 100);
    }

    #[tokio::test]
    async fn unknown_params_are_ignored() {
        let body = get("/setup/eureka_info?params=name,does_not_exist").await;
        let object = body.as_object().unwrap();
        assert_eq!(object.len(), 1);
        assert_eq!(body["name"], "Living Room");
    }

    #[test]
    fn cloud_device_id_uses_clean_udn_or_md5() {
        // 32 alnum after stripping dashes → used directly (uppercased).
        assert_eq!(
            cloud_device_id("12345678-1234-1234-1234-123456789abc"),
            "12345678123412341234123456789ABC"
        );
        // Otherwise MD5 hex uppercase (32 chars).
        let hashed = cloud_device_id("not-a-uuid");
        assert_eq!(hashed.len(), 32);
    }
}
